checkoff

```diff
@@ -0,0 +1,152 @@
+#!/usr/bin/env python3
+import argparse
+import json
+import subprocess
+import sys
+
+from ocflib.account.search import user_exists
+
+
+# Format of JSON file should be:
+# {
+#     "roster": [<list of OCF usernames of students>],
+#     "users": {
+#         "username1": [<list of labs checked off>],
+#         ...
+#     },
+#     "labs": {
+#         "a1": [<list of usernames checked off],
+#         ...
+#     },
+# }
+
+DATA_PATH = 'checkoff_data.json'
+
+
+def checkoff(args):
+    user = args.user
+    lab = args.lab
+    _check_lab_id(lab)
+
+    data = _load_data()
+    roster = data['roster']
+    users = data['users']
+    labs = data['labs']
+
+    if user not in roster:
+        print('{} not in class roster'.format(user))
+        return 1
+
+    if user in users:
+        if lab in users[user]:
+            print('{} already checked off for lab {}'.format(user, lab))
+            return 1
+
+        users[user].append(lab)
+    else:
+        users[user] = [lab]
+
+    if lab in labs:
+        labs[lab].append(user)
+    else:
+        labs[lab] = [user]
+
+    ans = input('OK to check off {} for lab {}? [y/N]\n'.format(user, lab))
+
+    if not ans or ans[0].lower() != 'y':
+        print('Cancelling checkoff')
+        return 1
+
+    _write_data(data)
+    _mail_checkoff(user, lab)
+
+    print('{} has been checked off for lab {}'.format(user, lab))
+
+
+def view(args):
+    user = args.user
+
+    data = _load_data()
+    roster = data['roster']
+    users = data['users']
+
+    if user not in roster:
+        print('{} not in class roster'.format(user))
+        return 1
+
+    completed_labs = []
+    if user in users:
+        completed_labs = sorted(users[user])
+
+    print('{} checked off for labs: {}'.format(user, completed_labs))
+
+
+def list_labs(args):
+    lab = args.lab
+    _check_lab_id(lab)
+
+    data = _load_data()
+    labs = data['labs']
+
+    if lab in labs:
+        for student in sorted(labs[lab]):
+            print(student)
+
+
+def _mail_checkoff(user, lab):
+    # subprocess needs input in bytes
+    msg = 'Congratulations {}, you have been checked off for lab {}'.format(user, lab).encode()
+    subject = 'Unix Sysadmin Decal Checkoff for Lab {}'.format(lab)
+    recipient = '{}@ocf.berkeley.edu'.format(user)
+    cc = 'decal+checkoffs@ocf.berkeley.edu'
+    subprocess.run(['mail', '-s', subject, '-c', cc, recipient], stdout=subprocess.PIPE, input=msg)
+
+
+def _load_data():
+    with open(DATA_PATH, 'r') as f:
+        data = json.load(f)
+    return data
+
+
+def _write_data(data):
+    with open(DATA_PATH, 'w') as f:
+        json.dump(data, f)
```

Comment:
I think you want a lock file or something in here so that if two people get checked off at the same time the second write doesn't overwrite the first write's data:

    read 1 -> data 1
    read 2 -> data 2
    data 1 -> write 1
    data 2 -> write 2 (overwrites data 1, since it read before write 1)

Reply:
Was thinking about that but was feeling lazy. I guess it could happen often enough that we should account for it

Reply:
Maybe just have `checkoff` acquire and release a lock? Since the read+write has to be atomic.

Reply:
I feel like it could happen a few times at least, although I doubt it'll happen often. The problem is more that we wouldn't notice it if it did happen. I think acquiring and releasing a lock would work well for the section where it writes anyway.

Reply:
That wont work, since the file could be changed by someone else between the calls to `_load_data` at line 31 and `_write data` at line 60 (this isn't such a far-fetched scenario, since the script waits for user input during this time). The `_write_data` function should be an append operation that reads the data file, modifies it, and then writes it.

Reply:
Yeah, so we can have the whole `checkoff` function (which reads, modifies, then writes) acquire and release a lock.